packages/seniman/seniman.d.ts

```diff
@@ -0,0 +1,145 @@
+declare namespace JSX {
+    interface Element {}
+    interface IntrinsicElements {
+        [elemName: string]: any;
+    }
+}
+
+declare module "seniman" {
+    export interface Component<T> {
+        id: string;
+        props: T;
+    }
+
+    export interface History {
+        pushState: (hrefString: string) => void; 
+        replaceState: (hrefString: string) => void; 
+    }
+
+    export interface Location {
+        host: string;
+        hostname: string;
+        origin: string;
+        protocol: string;
+        port: string;
+        href: () => null;
+        pathname: () => null;
+        search: () => null;
+        searchParams: () => null;
+        setHref: (hrefString: any) => void;
+    }
+
+    export interface ClientContext {
+        viewportSize: number; 
+        cookie: (cookieKey: string) => string; 
+        setCookie: (
+            cookieKey: string,
+            cookieValue: string,
+            expirationTime?: Date
+        ) => void;
+        history: History; 
+        location: Location; 
+        path: () => string;
+        navigate: (href: string) => void;
+        exec: (clientFnSpec: { clientFnId: number; serverBindFns: Function[] }) => void;
+    }
+    
+    export interface Window {
+        id: string;
+        href: string;
+        viewportSize: number;
+        readOffset: number;
+        cookieString: string;
+        client: ClientContext;
+    }
+
+    export interface Block {
+        id: string;
+    }
+
+    export interface Sequence {}
+
+    export function createRoot<T>(callback: Function): Root;
+
+    export interface Root {
+        hasWindow(windowId: string): boolean;
+        setRateLimit({ disabled: boolean }): void;
+        setDisableHtmlCompression(): void;
+        applyNewConnections(
+            ws: WebSocket,
+            options: { url: string; headers: Headers; ipAddress: string }
+        ): void;
+        initWindow(
+            ws: WebSocket,
+            options: {
+                windowId: string;
+                href: string;
+                viewportSize: number;
+                readOffset: number;
+                cookieString: string;
+            }
+        ): void;
+        reconnectWindow(
+            ws: WebSocket,
+            options: {
+                windowId: string;
+                href: string;
+                viewportSize: number;
+                readOffset: number;
+                cookieString: string;
+            }
+        ): void;
+        setServer(server: any): void;
+        getHtmlResponse(options: {
+            url: string;
+            headers: Headers;
+            ipAddress: string;
+            isSecure: boolean;
+        }): Response;
+        renderHtml(req: Request): Promise<any>;
```

Comment:
better to use `Promise<Response> `

Reply:
I think the `Root` object doesn't need external typing for now? It's now just used to be passed directly to the `serve` function in the user-side.

We might expose various external option settings through its methods in the future but nothing set at the moment yet, which we will then type, but the design's not there yet.

I would say the main Seniman packages' top-level exports would be the ones most helpful to type, which I think you've done.

Is the types file as it is now usable in a TypeScript project already?

Reply:
> I think the Root object doesn't need external typing for now? It's now just used to be passed directly to the serve function in the user-side.

make sense.

> Is the types file as it is now usable in a TypeScript project already?

Yes, probably I'll do another PR the example app?